validation/src/func.rs

```diff
@@ -0,0 +1,1236 @@
+        if self.is_any() || other.is_any() {
+            true
+        } else {
+            self.value_type() == other.value_type()
+        }
+    }
+}
+
+impl PartialEq<ValueType> for StackValueType {
+    fn eq(&self, other: &ValueType) -> bool {
+        if self.is_any() {
+            true
+        } else {
+            self.value_type() == *other
+        }
+    }
+}
+
+impl PartialEq<StackValueType> for ValueType {
+    fn eq(&self, other: &StackValueType) -> bool {
+        other == self
+    }
+}
+
+pub fn drive<T: FuncValidator>(
+    module: &ModuleContext,
+    func: &Func,
+    body: &FuncBody,
+) -> Result<T::Output, Error> {
+    let (params, result_ty) = module.require_function_type(func.type_ref())?;
+
+    let code = body.code().elements();
+    let code_len = code.len();
+    if code_len == 0 {
+        return Err(Error("Non-empty function body expected".into()));
+    }
+
+    let mut context = FunctionValidationContext::new(
+        &module,
+        Locals::new(params, body.locals())?,
+        DEFAULT_VALUE_STACK_LIMIT,
+        DEFAULT_FRAME_STACK_LIMIT,
+        result_ty,
+    );
+
+    let mut validator = T::new(&context, body);
+
+    for (position, instruction) in code.iter().enumerate() {
+        validator
+            .next_instruction(&mut context, instruction)
+            .map_err(|err| {
+                Error(format!(
+                    "At instruction {:?}(@{}): {}",
+                    instruction, position, err
+                ))
+            })?;
+    }
+
+    // The last `end` opcode should pop last instruction.
+    // parity-wasm ensures that there is always `End` opcode at
+    // the end of the function body.
+    assert!(context.frame_stack.is_empty());
+
+    Ok(validator.finish())
+}
+
+/// Function validation context.
+pub struct FunctionValidationContext<'a> {
+    /// Wasm module
+    pub module: &'a ModuleContext,
+    /// Local variables.
+    pub locals: Locals<'a>,
+    /// Value stack.
+    pub value_stack: StackWithLimit<StackValueType>,
+    /// Frame stack.
+    pub frame_stack: StackWithLimit<BlockFrame>,
+    /// Function return type.
+    pub return_type: BlockType,
+}
+
+impl<'a> FunctionValidationContext<'a> {
+    fn new(
+        module: &'a ModuleContext,
+        locals: Locals<'a>,
+        value_stack_limit: usize,
+        frame_stack_limit: usize,
+        return_type: BlockType,
+    ) -> Self {
+        let mut ctx = FunctionValidationContext {
+            module: module,
+            locals: locals,
+            value_stack: StackWithLimit::with_limit(value_stack_limit),
+            frame_stack: StackWithLimit::with_limit(frame_stack_limit),
+            return_type: return_type,
+        };
+        // push_label overflows only if the frame stack overflow happens. We assume that
+        // the limit is greater than 0 here. We don't expect `push_label` getting any other
+        // new errors.
+        let _ = push_label(
```

Comment:
why not add an expect then?

Reply:
You're right. As far as I remember there was some genius idea why it is that way, but now I doubt that it is worth it.

Although maybe we would want to consider to add a note on setting crazy limits like 1. : )